File: backend/app/schemas/reports.py
"""报告域 Pydantic Schema。

对齐 LLD §6.2 / §8.2 与前端 ``frontend/src/services/api/types.ts``：
- ``ReportResponse`` 支撑 ``GET /runs/{id}/report`` 与 ``GET /reports/{run_id}``；
  M2-7 起在原八字段之上增 ``outline`` / ``blocks`` 结构化超集（draft 旧报告为空数组）。
- ``ReportCitation`` 为 block 内联三字段引用；``ReportCitationItem`` 为
  ``GET /reports/{run_id}/citations`` 信源索引元素（关联键 + 九展示字段）。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

ReportStatus = Literal["draft", "final", "superseded"]
# 结构化区块四值（前端 ReportBlockType）
ReportBlockType = Literal["conclusion", "evidence", "dispute", "limitation"]
# 论断置信度三值
ClaimConfidence = Literal["single_source", "cross_verified", "inferred"]


class ReportCitation(BaseModel):
    """区块内联引用（数据点级溯源依据：角标 + 证据 + 原文片段）。"""

    evidence_id: str
    marker: str
    snippet: str


class ReportOutlineItem(BaseModel):
    """报告目录项（M2-7 语义 id：sec-overview/sec-findings/sec-disputes/sec-limitations）。"""

    id: str
    title: str
    type: str


class ReportBlock(BaseModel):
    """结构化区块（终稿渲染的唯一允许形态，对齐前端 ReportBlock）。"""

    id: str
    type: ReportBlockType
    text: str
    # conclusion/dispute 块带论断标识；evidence/limitation 块不带
    claim_id: str | None = None
    # evidence/limitation 块无置信度（与 fixture 形态一致）
    confidence: ClaimConfidence | None = None
    citations: list[ReportCitation] = []
    # 仅 dispute 块携带，关联冲突实体
    conflict_id: str | None = None


class ReportCitationItem(ReportCitation):
    """信源索引元素：内联三字段 + 回溯所需展示字段。

    线上共 10 字段：关联键 ``evidence_id`` 加九展示字段（marker/snippet/url/
    title/domain/source_type/source_level/credibility/published_at）；
    除 published_at 可空外，M2-6 后其余字段实际总有值，可选声明以前端冻结形态为准。
    """

    url: str
    title: str
    domain: str | None = None
    source_type: str | None = None
    source_level: str | None = None
    credibility: str | None = None
    published_at: datetime | None = None


class ReportResponse(BaseModel):
    """报告视图：Markdown 正文 + M2-7 结构化终稿超集。"""

    id: str
    run_id: str
    template_id: str
    status: ReportStatus = "draft"
    content_md: str
    token_used: int = 0
    created_at: datetime
    updated_at: datetime
    # M2-7：final 终稿从 content_json 带出；draft/历史旧报告为空数组
    outline: list[ReportOutlineItem] = []
    blocks: list[ReportBlock] = []

    model_config = ConfigDict(from_attributes=True)


__all__ = [
    "ClaimConfidence",
    "ReportBlock",
    "ReportBlockType",
    "ReportCitation",
    "ReportCitationItem",
    "ReportOutlineItem",
    "ReportResponse",
    "ReportStatus",
]

File: frontend/src/services/api/reports.ts
// 报告 API（M1：GET /runs/{run_id}/report、GET /reports/{report_id}；
// M2-7：GET /reports/{report_id}/citations 数据点溯源索引）
import { http } from '../http/http'
import type { ReportCitationItem, ReportResponse, StructuredReportResponse } from './types'

// 按运行获取报告（指挥舱终态后进入报告页使用）
export function getRunReport(runId: string): Promise<ReportResponse> {
  return http<ReportResponse>(`/runs/${encodeURIComponent(runId)}/report`)
}

// 按运行获取报告（契约 GET /reports/{run_id}：M1 路径参数语义同 run_id；
// 报告未生成时后端返回 422 validation_error。M4 起该参数才作为独立 report id 使用）
export function getReport(runId: string): Promise<ReportResponse> {
  return http<ReportResponse>(`/reports/${encodeURIComponent(runId)}`)
}

// 获取结构化终稿（WP-16 blocks 轨道）。
// M2-7 契约冻结前与 getReport 同路径：mock 对结构化剧本返回 markdown+blocks 超集；
// 真实后端当前仅回 markdown（无 blocks 字段，页面据此走 markdown 轨道）。
// M2-7 冻结后若路径或形态调整，只改本函数与 StructuredReportResponse 类型。
export function getStructuredReport(runId: string): Promise<StructuredReportResponse> {
  return http<StructuredReportResponse>(`/reports/${encodeURIComponent(runId)}`)
}

// 获取报告引用列表（数据点级溯源：角标 ↔ 证据 URL/原文片段，M2-7 冻结）
export function getReportCitations(reportId: string): Promise<ReportCitationItem[]> {
  return http<ReportCitationItem[]>(
    `/reports/${encodeURIComponent(reportId)}/citations`
  )
}
